text/0096-rawkv-stale-read.md

```diff
@@ -0,0 +1,97 @@
+
+    <p align="center">
+        <img src="../media/txnkv-stale-read.svg" width="450px"/>
+    </p>
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before (read committed).
+
+The follower will read the local storage with a `read_ts` that reuses the mechanism from the stale read of TxnKV. Surprisingly, the leader can also stale read, and in this case, the leader can read locally without a lease.
+
+Note that the freshness of data may regress from the point of view of the client since the client may choose different followers to read from time to time.
+
+<p align="center">
+    <img src="../media/rawkv-stale-read.svg" width="450px"/>
+</p>
+
+
+## Detailed Design
+
+This RFC proposes stale read on RawKV which allows reading stale data from the not-leader replica with a lossy time-bound. Unlike the stale read on TxnKV, the stale read on RawKV breaks its original linear consistency guarantee -- downgrades to eventually consistency. TiKV only guarantees that the stale data read from followers are written before.
+
+### TiKV
+
+ While trying to read data, clients should specify a timestamp which attachs to the request header as `read_ts`, typically a timestamp few seconds ago. The replica should read the local storage with the `read_ts` and reuses the mechanism from the stale read of TxnKV. This requires the replica to check the `read_ts` against the `safe_ts` which is advaneced by `CheckLeader` message from the store of the leader (for follower) or `resolve-ts` worker (for leader). As long as the `safe_ts` is no less than `read_ts`, the replica is allowed to read the key from local storage. Notice that there is no lock for the RawKV regions, thus the `resolve-ts` worker advanced the `safe_ts` by requesting the TSO for the latest timestamp.
+
+### Client
+
+The `read_ts` specified by the client could be acquired by the following ways:
+
+1. Calculate a timestamp from the physical time from the local. The `read_ts` might suffer from the clock drift and exceed the max timestamp allocated from TSO. The client will fail to read any data even if that target replica is the leader since the `safe_ts` of the replica don't catch up with the `read_ts`. **Deploying NTP services** in the cluster might mitigate this issue.
```

Comment:
if `read_ts` exceeds the max timestamp allocated from TSO, maybe we can just return the latest data instead of no data.

Reply:
Then the `read_ts` will lost its restriction to the data freshness since some very stale replicas might be chosen.